### DISCUSSION TITLE
Building a Quarto Website that Consolidates Multiple GitHub Repositories

### DISCUSSION THREAD

**Discussion by fredguth** [quarto-dev/quarto-cli#9399] [Q&A] [answered]
### Description

Has anyone created or come across a Quarto website that integrates and renders content from multiple GitHub repositories? 

I'm interested in developing a journal-style website where each article is managed in its own separate GitHub repository. Specifically, I'm looking for examples or advice on how to consolidate these repositories into a single Quarto website, akin to how a journal might publish articles from various authors. 

Although I'm uncertain whether this  more to a Git question than a Quarto one, any guidance, examples, or references to similar projects would be appreciated.

**Comment by mcanouil** [comment, marked as answer]
Quarto cannot do that.
It's all about your "file system".

When working with Git repositories, the Git way to embed multiple repositories into one is Git Submodule.
On GitHub, you can set Dependabot to automatically make PRs to update your submodules (https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#package-ecosystem).
You can also set some workflow triggered by merged PRs to render with Quarto.


Another approach would be to create a GitHub App to clone the repositories at render time.
It's a bit the idea or "R Universe".

**Reply by fredguth** [reply]
Thanks for the prompt answer.  I figured it was more of a git question.  I will look into the `dependabot` you mentioned.  I am not very proficient with Github flows. Have you seen such approach in a large quarto website?

**Reply by mcanouil** [reply]
Not seen this and I would say it's unlikely because it's not straightforward.
A lot of Git users try their best to never have to use Git Submodules 😅
Dependabot is the only reason why I suggested it in the first place because it simplifies a lot the process to update submodules.

**Reply by cscheid** [reply]
@mjskay sorry to ping you at random here, but do you have any insights on this with your experience at JoVI?

**Reply by mjskay** [reply]
Happy to be pinged at random for stuff like this :).

At [jovi](https://www.journalovi.org/) I initially tried doing this using an overall quarto site with subfolders for papers as git submodules and found it to be kind of a pain. We eventually abandoned this for a few reasons, some of which I might be misremembering (or which might have better solutions):
- rendering a paper individually (as authors might) seemed to conflict with rendering the whole site at times. I can't remember exactly the issues, but I think it was things like if you just checkout the paper repo and render it you could get a slightly different-looking paper than if you render the whole site from the root, depending on theme settings and whatnot.
- we also decided to allow non-quarto papers, and supporting that seemed potentially more annoying
- anticipating future problems: even with caching I can imagine a future scenario where rendering the whole site becomes a pain with lots of papers (I'm less certain of this).

We also just found a simpler (at least in the short-term) solution using Github pages:
- create a github organization
- use a Github pages-enabled organization site repo as the main journal site; e.g. [this repo](https://github.com/journalovi/journalovi.github.io) renders the main site under `/` at [journalovi.org](https://www.journalovi.org/)
- for each paper, create a Github pages-enabled repo in the organization using a standard naming scheme, like `2023-author_name-title_stub`. This will be rendered to `/2023-author_name-title_stub/` at [journalovi.org](https://www.journalovi.org/); e.g. [this repo](https://github.com/journalovi/2023-park-gatherplots) renders [this article](https://www.journalovi.org/2023-park-gatherplots/) on journalovi.org
- Finally, if we want to get a list of papers on the main site, we can do this using the Github API during main site render. For example, [this page](https://www.journalovi.org/under-review.html) lists all papers currently under review. It does this by finding all repos under the Github organization with the `article` topic applied to it (topics are basically arbitrary tags you can assign repos on Github). It also does a bit of categorization, e.g. papers with the `article` and `vixai` topic are put into a separate list. You can see the code for this [here](https://github.com/journalovi/journalovi.github.io/blob/main/under-review.qmd). It is currently a little experimental and probably would be good to solidify into a package once we iron it out (unless this kind of functionality moves into Quarto at some point... ;) ).

**Reply by fredguth** [reply]
Very interesting.  This is very similar to what I want to accomplish.  Thanks a lot!

**Reply by fredguth** [reply]
Also.. total coincidence, PAC Learning Framework (subject of the article under review) was an important topic of my Master's Dissertation.
